pkg/git/git.go

```diff
@@ -84,53 +86,99 @@ func Fetch(logger *zap.SugaredLogger, spec FetchSpec) error {
 		return err
 	}
 
-	fetchArgs := []string{"fetch", "--recurse-submodules=yes"}
+	fetchArgs := []string{"fetch"}
+	if spec.Submodules {
+		fetchArgs = append(fetchArgs, "--recurse-submodules=yes")
+	}
 	if spec.Depth > 0 {
 		fetchArgs = append(fetchArgs, fmt.Sprintf("--depth=%d", spec.Depth))
 	}
-	fetchArgs = append(fetchArgs, "origin", spec.Revision)
 
-	if _, err := run(logger, "", fetchArgs...); err != nil {
-		// Fetch can fail if an old commitid was used so try git pull, performing regardless of error
-		// as no guarantee that the same error is returned by all git servers gitlab, github etc...
-		if _, err := run(logger, "", "pull", "--recurse-submodules=yes", "origin"); err != nil {
-			logger.Warnf("Failed to pull origin : %s", err)
-		}
-		if _, err := run(logger, "", "checkout", spec.Revision); err != nil {
+	// Fetch the revision and verify with FETCH_HEAD
+	fetchParam := []string{spec.Revision}
+	checkoutParam := "FETCH_HEAD"
+
+	if spec.Refspec != "" {
+		// if refspec is specified, fetch the refspec and verify with provided revision
+		fetchParam = strings.Split(spec.Refspec, " ")
+		checkoutParam = spec.Revision
+	}
+
+	// git-init always creates and checks out an empty master branch. When the user requests
+	// "master" as the revision, git-fetch will refuse to update the HEAD of the branch it is
+	// currently on. The --update-head-ok parameter tells git-fetch that it is ok to update
+	// the current (empty) HEAD on initial fetch.
+	// The --force parameter tells git-fetch that its ok to update an existing HEAD in a
+	// non-fast-forward manner (though this cannot be possible on initial fetch, it can help
+	// when the refspec specifies the same destination twice)
+	fetchArgs = append(fetchArgs, "origin", "--update-head-ok", "--force")
```

Comment:
I don't see the value in putting an additional conditional for that case. This is (guaranteed to be) the first fetch command run after git-init, so its implied that 
a. it will always be the master branch, unless git-init changes behavior 
b. we are not expecting any unintended updates to existing refs